build(aio): auto-link more code items

petebacondarwin opened the pull request:
I looked at various ways to achieve this but it is not possible to know until after
rendering exactly what code is going to be on the page.
So I opted for post-processing the HTML, effectively expanding what we were doing
for simple code elements already to parse through the text of all `code`, `code-example`
and `code-pane` elements for potentially matching code items.
This means that we could accidentally get a false positive but a quick anecdotal survey
of the docs hasn't found any so far.

mary-poppins commented:
You can preview 52599ff at https://pr19109-52599ff.ngbuilds.io/.

gkalpak commented on aio/tools/transforms/angular-base-package/post-processors/auto-link-code.spec.js:
```diff
@@ -1,7 +1,7 @@
 var Dgeni = require('dgeni');
 
-describe('autoLinkCode post-processor', () => {
+fdescribe('autoLinkCode post-processor', () => {
```
😱

gkalpak commented on aio/tools/transforms/angular-base-package/post-processors/auto-link-code.js:
```diff
@@ -3,34 +3,79 @@ const is = require('hast-util-is-element');
+                parent.children.splice(index, 1, createLinkNode(docs[0], node.value));
+              } else {
+                // Parse the text for words that we can convert to links
+                const nodes = textContent(node).split(/([A-Za-z0-9]+)/)
```
OOC, do we need `textContent()` here (given we are visiting `text` nodes)?

##
AFAICT, `.split(/[^A-Za-z0-9]+/)` would work too (and is marginally more straight forward :grin:).
BTW, I would expect including `_` too. Was is left out on purpose?

petebacondarwin replied:
@gkalpak your regex doesn't work because you then lose the whitespace:

```
> 'a b c'.split(/([a-z])/)
[ '', 'a', ' ', 'b', ' ', 'c', '' ]
> 'a b c'.split(/[^a-z]/)
[ 'a', 'b', 'c' ]
```

gkalpak replied:
Yeah, I thought you didn't need them. I know realize you need all "tokens" :grin:

gkalpak commented on aio/tools/transforms/angular-base-package/post-processors/auto-link-code.js:
```diff
@@ -3,34 +3,79 @@ const is = require('hast-util-is-element');
+                      nodes.push(createLinkNode(docs[0], word));
+                    } else {
+                      // this is just text so either attach it to the previous text node or create a new one
+                      const previous = nodes[node.length];
```
Why `nodes[node.length]`??? What do I miss?

petebacondarwin replied:
oops, `length-1` I think.

petebacondarwin replied:
I realise that we don't  actually need to join up these text nodes as at the end of the postProcessing we just render the AST back to an HTML string. So we can just push new text nodes.

gkalpak reviewed (changes requested):
OMG!!! This is huge 💯 💯 💯 

One thing: Could we update the styling of hovered links inside `code-example`/`code-pane` elements (e.g. underline).

petebacondarwin commented:
I'll add some new CSS but that is venturing into wild places I fear to tread!

petebacondarwin committed:
build(aio): auto-link more code items

We now parse all code blocks, after they have been rendered by dgeni
and insert links to API docs that match "words" in the code.

gkalpak commented on aio/tools/transforms/angular-base-package/post-processors/auto-link-code.spec.js:
```diff
@@ -44,25 +44,34 @@ fdescribe('autoLinkCode post-processor', () => {
-    const doc = { docType: 'test-doc', renderedContent: '<code>MyClass() MyClass+2</code>' };
+    aliasMap.addDoc({ docType: 'function', id: 'myFunc', aliases: ['myFunc'], path: 'ng/myfunc' });
+    aliasMap.addDoc({ docType: 'const', id: 'MY_CONST', aliases: ['MY_CONST'], path: 'ng/my_const' });
+    const doc = { docType: 'test-doc', renderedContent: '<code>myFunc() {\n  return new MyClass(MY_CONST));\n}</code>' };
```
Extra closing `)` 😛

mary-poppins commented:
You can preview 53897a8 at https://pr19109-53897a8.ngbuilds.io/.

mary-poppins commented:
You can preview eac48ac at https://pr19109-eac48ac.ngbuilds.io/.

matsko commented:
Landed as 9ab9437319987a251e1b1104402acd0cc474a907 and 9d93c859d7791fb16c216ff95265c58ba1000c5a
